File: crates/merchant-core/src/validation.rs
use thiserror::Error;
use crate::{Competitor, EvidenceSource};

#[derive(Clone, Debug, Error, PartialEq, Eq)]
pub enum ValidationError {
    #[error("Project name is required")]
    MissingProjectName,
    #[error("Research objective is required")]
    MissingObjective,
    #[error("Currency must be exactly three uppercase ASCII letters")]
    InvalidCurrency,
    #[error("Invalid evidence source: {0}")]
    InvalidEvidence(String),
}

pub fn validate_evidence_sources(sources: &[EvidenceSource]) -> Result<(), ValidationError> {
    let mut ids = std::collections::HashSet::new();
    for source in sources {
        if source.schema_version != crate::SCHEMA_VERSION {
            return Err(ValidationError::InvalidEvidence(format!("{} has an unsupported schema version", source.id)));
        }
        if !source.id.starts_with("S-") || source.id.len() <= 2 || !ids.insert(&source.id) {
            return Err(ValidationError::InvalidEvidence("source IDs must be unique and start with S-".into()));
        }
        let valid_url = ["https://", "http://"].iter().any(|prefix| {
            source.url.strip_prefix(prefix).is_some_and(|rest| !rest.trim().is_empty())
        });
        if !valid_url {
            return Err(ValidationError::InvalidEvidence(format!("{} needs an http or https URL", source.id)));
        }
        if source.title.trim().is_empty() {
            return Err(ValidationError::InvalidEvidence(format!("{} needs a title", source.id)));
        }
    }
    Ok(())
}

pub fn validate_competitors(competitors: &[Competitor], project_currency: &str) -> Result<(), ValidationError> {
    let mut ids = std::collections::HashSet::new();
    for competitor in competitors {
        if competitor.schema_version != crate::SCHEMA_VERSION {
            return Err(ValidationError::InvalidEvidence(format!("{} has an unsupported schema version", competitor.id)));
        }
        if !competitor.id.starts_with("C-") || competitor.id.len() <= 2 || !ids.insert(&competitor.id) {
            return Err(ValidationError::InvalidEvidence("competitor IDs must be unique and start with C-".into()));
        }
        if competitor.product.trim().is_empty() || competitor.currency != project_currency {
            return Err(ValidationError::InvalidEvidence(format!("{} has an invalid product or currency", competitor.id)));
        }
        if competitor.price.is_some_and(|price| price.decimal().is_sign_negative()) {
            return Err(ValidationError::InvalidEvidence(format!("{} has a negative price", competitor.id)));
        }
    }
    Ok(())
}

pub fn validate_project_name(value: &str) -> Result<(), ValidationError> {
    if value.trim().is_empty() {
        return Err(ValidationError::MissingProjectName);
    }
    Ok(())
}

pub fn validate_objective(value: &str) -> Result<(), ValidationError> {
    if value.trim().is_empty() {
        return Err(ValidationError::MissingObjective);
    }
    Ok(())
}

pub fn validate_currency(value: &str) -> Result<(), ValidationError> {
    if value.len() == 3 && value.bytes().all(|byte| byte.is_ascii_uppercase()) {
        return Ok(());
    }
    Err(ValidationError::InvalidCurrency)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validates_required_project_fields_and_iso_currency_shape() {
        assert!(validate_project_name("Mechanical Keyboards India").is_ok());
        assert!(validate_objective("Assess demand").is_ok());
        assert!(validate_currency("INR").is_ok());
        assert!(validate_currency("inr").is_err());
        assert!(validate_currency("INRR").is_err());
    }
}

File: packages/core/src/validation.ts
import type {
  Competitor,
  CostAssumptions,
  EvidenceSource,
} from "@open-merchant/shared";
import { competitorIdSchema, currencyCodeSchema, evidenceSourceIdSchema } from "@open-merchant/shared";

import { Decimal, parseAmount } from "./money";

/**
 * Workspace validation rules, ported from the legacy Rust engine: malformed
 * or out-of-range data is rejected with a specific reason — never silently
 * repaired. Every check throws on the first failure found.
 */

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

const HTTP_PREFIXES = ["https://", "http://"] as const;
const ZERO = new Decimal(0);
const HUNDRED = new Decimal(100);

function fail(message: string): never {
  throw new ValidationError(message);
}

function amountOf(raw: string): Decimal {
  try {
    return parseAmount(raw);
  } catch {
    return fail(`Invalid decimal amount: ${JSON.stringify(raw)}`);
  }
}

export function validateProjectName(value: string): void {
  if (value.trim().length === 0) fail("Project name is required");
}

export function validateObjective(value: string): void {
  if (value.trim().length === 0) fail("Research objective is required");
}

export function validateCurrency(value: string): void {
  if (!currencyCodeSchema.safeParse(value).success) {
    fail("Currency must be exactly three uppercase ASCII letters");
  }
}

/** Evidence sources must be unique, http(s), titled, and well-formed. */
export function validateEvidenceSources(sources: EvidenceSource[]): void {
  const seen = new Set<string>();
  for (const source of sources) {
    if (!evidenceSourceIdSchema.safeParse(source.id).success || seen.has(source.id)) {
      fail("source IDs must be unique and start with S-");
    }
    seen.add(source.id);

    const validUrl = HTTP_PREFIXES.some(
      (prefix) => source.url.startsWith(prefix) && source.url.slice(prefix.length).trim().length > 0,
    );
    if (!validUrl) fail(`${source.id} needs an http or https URL`);
    if (source.title.trim().length === 0) fail(`${source.id} needs a title`);
  }
}

export function validateCompetitors(competitors: Competitor[], projectCurrency: string): void {
  const seen = new Set<string>();
  for (const competitor of competitors) {
    if (!competitorIdSchema.safeParse(competitor.id).success || seen.has(competitor.id)) {
      fail("competitor IDs must be unique and start with C-");
    }
    seen.add(competitor.id);

    if (competitor.product.trim().length === 0 || competitor.currency !== projectCurrency) {
      fail(`${competitor.id} has an invalid product or currency`);
    }
    if (competitor.price !== null && amountOf(competitor.price).isNegative()) {
      fail(`${competitor.id} has a negative price`);
    }
  }
}

export function validateAssumptions(assumptions: CostAssumptions, projectCurrency: string): void {
  if (assumptions.currency !== projectCurrency) {
    fail("Assumptions must use the project schema and currency");
  }

  for (const cost of [assumptions.acquisitionCost, assumptions.shippingCost, assumptions.otherCosts]) {
    if (amountOf(cost).isNegative()) fail("Costs cannot be negative");
  }

  for (const rate of [assumptions.marketplaceFeeRate, assumptions.paymentFeeRate]) {
    const value = amountOf(rate);
    if (value.isNegative() || value.greaterThan(HUNDRED)) {
      fail("Fee rates must be between 0 and 100");
    }
  }

  for (const price of [
    assumptions.scenarioPrices.low,
    assumptions.scenarioPrices.base,
    assumptions.scenarioPrices.high,
  ]) {
    if (price !== null && amountOf(price).lessThanOrEqualTo(ZERO)) {
      fail("Scenario prices must be positive");
    }
  }
}
